From 512a4a089498a3c0c393b99b4a24bbdc1355daea Mon Sep 17 00:00:00 2001
From: Zhe Zhang <zhz@apache.org>
Date: Fri, 8 Jan 2016 13:14:36 -0800
Subject: [PATCH 1079/2748] HDFS-9626.
 TestBlockReplacement#testBlockReplacement fails
 occasionally. Contributed by Xiao Chen.

(cherry picked from commit 40dabae236c82881e28614fbe1fcd1cf707b7033)

Change-Id: I0be4745f7bcad0e933d8784cd51d52e3932cd287
---
 .../hdfs/server/datanode/TestBlockReplacement.java |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReplacement.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReplacement.java
index 6e4a19b..82145e9 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReplacement.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/datanode/TestBlockReplacement.java
@@ -197,9 +197,10 @@ public void testBlockReplacement() throws Exception {
       LOG.info("Testcase 4: invalid del hint " + proxies.get(0) );
       assertTrue(replaceBlock(b, proxies.get(0), proxies.get(1), source));
       // after cluster has time to resolve the over-replication,
-      // block locations should contain two proxies,
-      // and either source or newNode, but not both.
-      checkBlocks(proxies.toArray(new DatanodeInfo[proxies.size()]), 
+      // block locations should contain any 3 of the blocks, since after the
+      // deletion the number of racks is still >=2 for sure.
+      // See HDFS-9314 for details, espacially the comment on 18/Nov/15 14:09.
+      checkBlocks(new DatanodeInfo[]{},
           fileName.toString(), 
           DEFAULT_BLOCK_SIZE, REPLICATION_FACTOR, client);
     } finally {
-- 
1.7.9.5

